wandb/sdk/internal/sender.py

```diff
@@ -1434,27 +1435,33 @@ def send_use_artifact(self, record: "Record") -> None:
             )
 
     def send_request_log_artifact(self, record: "Record") -> None:
-        assert record.control.req_resp
+        assert record.control.mailbox_slot
         result = proto_util._result_from_record(record)
         artifact = record.request.log_artifact.artifact
         history_step = record.request.log_artifact.history_step
 
         try:
-            res = self._send_artifact(artifact, history_step)
+            res, future = self._send_artifact(artifact, history_step)
             assert res, "Unable to send artifact"
             result.response.log_artifact_response.artifact_id = res["id"]
             logger.info(f"logged artifact {artifact.name} - {res}")
+            if future is not None:
+                # respond to the request only after the artifact is fully committed
+                future.add_done_callback(lambda _: self._respond_result(result))
+            else:
+                self._respond_result(result)
         except Exception as e:
             result.response.log_artifact_response.error_message = (
                 f'error logging artifact "{artifact.type}/{artifact.name}": {e}'
             )
 
-        self._respond_result(result)
-
     def send_artifact(self, record: "Record") -> None:
         artifact = record.artifact
         try:
-            res = self._send_artifact(artifact)
+            res, future = self._send_artifact(artifact)
+            # wait for future to complete in send artifact
+            if future is not None:
+                future.done()
```

Comment:
`done()` just checks whether the future is done, we want:
```suggestion
                future.result()
```

Reply:
fixed